components/builder-api-client/src/lib.rs

```diff
@@ -139,10 +142,45 @@ impl Client {
         let mut encoded = String::new();
         res.read_to_string(&mut encoded).map_err(Error::IO)?;
         debug!("Body: {:?}", encoded);
-        let rd: ReverseDependencies = serde_json::from_str(&encoded).map_err(Error::JSON)?;
+        let rd: ReverseDependencies = serde_json::from_str(&encoded).map_err(Error::Json)?;
         Ok(rd.rdeps.to_vec())
     }
 
+    /// Promote a job group to a channel
+    ///
+    /// # Failures
+    ///
+    /// * Remote API Server is not available
+    pub fn job_group_promote(
+        &self,
+        group_id: u64,
+        channel: &str,
+        token: &str,
+    ) -> Result<Vec<String>> {
+        debug!("Promoting job group {} to channel {}", group_id, channel);
+        let url = format!("jobs/group/{}/promote/{}", group_id, channel);
+        let res = self.add_authz(self.0.post(&url), token).send().map_err(
+            Error::HyperError,
+        )?;
+
+        if res.status != StatusCode::Ok {
+            return Err(err_from_response(res));
+        }
+
+        match decoded_response::<UnsuccessfulJobGroupPromote>(res)
```

Comment:
I'm happy to rename the struct if you think it would make things clearer, but that won't affect the error handling below.

By abstracting away the parsing of response bodies into `decoded_response`, we've lost the ability to have fine grained control over what to do with empty response bodies, since that method returns something that can be deserialized by serde.  Empty response bodies are perfectly normal for an HTTP API to have, but when you try and parse an empty response body with serde, you get an error, which we handle below.

I'm of the opinion that it's better to have your public facing HTTP API be correct and follow the principle of least surprise, and handle any oddities in your client libraries.  That's how this is setup right now.  We can change it so that the client code is more straightforward from an error handling perspective, but doing so will require making our HTTP API less intuitive, in my opinion.

In this particular case, there's just a wrinkle in how we handle success and failure, and I think it's best if it's not exposed to end users.